Lib/functools.py

```diff
@@ -273,43 +273,123 @@ def reduce(function, sequence, initial=_initial_missing):
 ### partial() argument application
 ################################################################################
 
+
+class __PlaceholderTypeBase:
+    """The type of the Placeholder singleton.
+
+    Used as a placeholder for partial arguments.
+    """
+    __instance = None
+
+    def __new__(cls):
+        if cls.__instance is None:
+            cls.__instance = object.__new__(cls)
+        return cls.__instance
+
+    def __bool__(self):
+        raise TypeError("Placeholder should not be used in a boolean context")
+
+    def __repr__(self):
+        return 'Placeholder'
+
+    def __reduce__(self):
+        return 'Placeholder'
+
+
+Placeholder = type('PlaceholderType', (__PlaceholderTypeBase,), {})()
```

Comment:
Nit: shouldn't we use `SCREAMING_SNAKE_CASE` for constant variables?

```suggestion
PLACEHOLDER = type('PlaceholderType', (__PlaceholderTypeBase,), {})()
```

I think sentinels in `dataclasses` would be a good example:

https://github.com/python/cpython/blob/ead676516d4687b5e6d947a5c0808b5b93ba87e3/Lib/dataclasses.py#L172-L190

------

Also, note that this implementation does not support serialization. Is it an intentional behavior?

```python
In [1]: import dataclasses
   ...: import functools
   ...: import pickle

In [2]: str(dataclasses.MISSING)  # low-hanging fruit for better representation
Out[2]: '<dataclasses._MISSING_TYPE object at 0x100cf5550>'

In [3]: repr(dataclasses.MISSING)  # low-hanging fruit for better representation
Out[3]: '<dataclasses._MISSING_TYPE object at 0x100cf5550>'

In [4]: type(dataclasses.MISSING)
Out[4]: <class 'dataclasses._MISSING_TYPE'>

In [5]: str(functools.Placeholder)
Out[5]: 'Placeholder'

In [6]: repr(functools.Placeholder)
Out[6]: 'Placeholder'

In [7]: type(functools.Placeholder)
Out[7]: <class 'functools.PlaceholderType'>

In [8]: pickle.loads(pickle.dumps(dataclasses.MISSING))
Out[8]: <dataclasses._MISSING_TYPE object at 0x102c58a50>

In [9]: pickle.loads(pickle.dumps(functools.Placeholder))
Unexpected exception formatting exception. Falling back to standard exception
Traceback (most recent call last):
  File "/Users/PanXuehai/Projects/cpython/venv/lib/python3.14/site-packages/IPython/core/interactiveshell.py", line 3577, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
    ~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<ipython-input-9-bdbe2f9368c5>", line 1, in <module>
    pickle.loads(pickle.dumps(functools.Placeholder))
                 ~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
_pickle.PicklingError: Can't pickle <class 'functools.PlaceholderType'>: attribute lookup PlaceholderType on functools failed
```

Reply:
Regarding naming, this follows a snake-case convention in line with `None` and similar given then nature of its implementation.

Regarding serialization error, I will look into it. It should not be this way.

Reply:
> Regarding naming, this follows a snake-case convention in line with `None` and similar given then nature of its implementation and purpose.

I think we should align with `dataclasses.{KW_ONLY,MISSING}` since `functools.PLACEHOLDER` is a public API rather than something like:

```python
class _Sentinel:
    __slots__ = ()

    def __repr__(self):
        return '<sentinel>'

_sentinel = _Sentinel()

def func(x=_sentinel):
    ...

class Foo:
    _marker = object()

    def bar(self, x=_marker):
        ...
```

> Regarding serialization error, I will look into it. It should not be this way.

`pickle` will serialize class/function with its `{obj.__module__}.{obj.__qualname__}`, then it can be imported in another process during deserialization. The pickle error is due to there being no member `PlaceholderType` in module `functools`.

Here is another object intentionally removed pickle support.

```python
In [1]: import sys

In [2]: sys.float_info
Out[2]: sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

In [3]: type(sys.float_info)
Out[3]: <class 'sys.float_info'>

In [4]: import pickle

In [5]: pickle.loads(pickle.dumps(sys.float_info))
Unexpected exception formatting exception. Falling back to standard exception
Traceback (most recent call last):
  File "/Users/PanXuehai/Projects/cpython/venv/lib/python3.14/site-packages/IPython/core/interactiveshell.py", line 3577, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
    ~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<ipython-input-5-fd1206149756>", line 1, in <module>
    pickle.loads(pickle.dumps(sys.float_info))
                 ~~~~~~~~~~~~^^^^^^^^^^^^^^^^
_pickle.PicklingError: Can't pickle <class 'sys.float_info'>: it's not the same object as sys.float_info
```

The `sys.float_info` object and its type `type(sys.float_info)` share the same identifier `sys.float_info`.

Reply:
> I think we should align with `dataclasses.{KW_ONLY,MISSING}` since `functools.PLACEHOLDER` is a public API rather than something like:

It aligns with `NotImplemented`, `None` and similar which are singletons implemented in C.

The focus is a C implementation. Python implementation is a fallback and what you currently see in Python is only to emulate how C implementation functions.

The error that you are seeing is for C implementation of it.

This has been thoroughly discussed before and the concept of it is not likely going to change.

However, pickling issue needs sorting out. The pickling error needs a different solution to be addressed. The case at hand is that of `None`. But `type(None).__module__ = 'builtins`, however `builtins.NoneType` does not exist.

Initially, `PlaceholderType` was exposed and all was functioning well, but it got messed up after decision to hide it in the same way as `NoneType` is hidden.

I am well aware how pickle works for simple cases, it is just that this case is a bit different. If you want to look at `object.c` where `NoneType` is implemented and help me figure out how pickling for it works, help is always appreciated.

Reply:
It is a bit of a pickle here...

Either:
a) Exceptions to `pickle` protocol need to be added (same way as for NotImplemented and None)
b) Expose `functools.PlaceholderType`.

I am leaning to b).

`Placeholder` constant is more significant than constant groups such as the ones in `inspect` or dataclass`, but I don't think it is significant enough (such as `None`) to deserve a special case in `pickle`.

Reply:
How about:

```python
def _get_placeholder():
    return __import__("functools").Placeholder

class __PlaceholderType:
    def __reduce__(self):
        return _get_placeholder, ()
```

Reply:
Btw, the core difference from dataclasses constants (and others) is it being a singleton. I.e.:
```
type(dataclasses.MISSING)() is dataclasses.MISSING   # False
type(Placeholder)() is Placeholder           # True
type(None)() is None    # True
```

Reply:
Serializing the object is not an issue. 

Python implementation works as it is. Add import error to `functools.py` and check:
```python
try:
    raise ImportError
    from _functools import partial, Placeholder
except ImportError:
    pass
```

Fixing it in C is easy. `__reduce__(self): return "Placeholder"` is the only addition needed.

The issue is serializing `type(Placeholder)`

Reply:
> The issue is serializing `type(Placeholder)`

I think there will be no real-world use case for `type(Placeholder)`. Just using `obj is Placeholder` is enough.

Reply:
It is more a matter of robustness.

E.g. I have never seen `type(None)() is None` used either.

But such design helps ensure uniqueness allowing for things like `isinstance(obj, (str, int, type(None))`.

In turn, there is a need for serializing `type(None)`.

So it is simply covering the bases for constant of certain significance.

Reply:
The only simple solution that I see is exposing `PlaceholderType`.

Adding exception to `pickle` can not be easily and sensibly done. It has complications beyond what is done for `NoneType`.

Reply:
> The issue is serializing `type(Placeholder)`

The relavent code is here:

https://github.com/python/cpython/blob/ead676516d4687b5e6d947a5c0808b5b93ba87e3/Lib/pickle.py#L1136-L1143

Here is how `NoneType` is pickled:

```python
In [1]: import pickle

In [2]: pickle.dumps(None)
Out[2]: b'\x80\x04N.'

In [3]: pickle.dumps(type(None))
Out[3]: b'\x80\x04\x95\x1a\x00\x00\x00\x00\x00\x00\x00\x8c\x08builtins\x94\x8c\x04type\x94\x93\x94N\x85\x94R\x94.'

In [4]: import pickletools

In [5]: list(pickletools.genops(pickle.dumps(None)))
Out[5]: 
[
│   (<pickletools.OpcodeInfo object at 0x1104656c0>, 4, 0),
│   (<pickletools.OpcodeInfo object at 0x110464520>, None, 2),
│   (<pickletools.OpcodeInfo object at 0x110465720>, None, 3)
]

In [6]: pickletools.dis(pickle.dumps(None))
    0: \x80 PROTO      4
    2: N    NONE
    3: .    STOP
highest protocol among opcodes = 2

In [7]: list(pickletools.genops(pickle.dumps(type(None))))
Out[7]: 
[
│   (<pickletools.OpcodeInfo object at 0x1104656c0>, 4, 0),
│   (<pickletools.OpcodeInfo object at 0x110465780>, 26, 2),
│   (<pickletools.OpcodeInfo object at 0x1104646a0>, 'builtins', 11),
│   (<pickletools.OpcodeInfo object at 0x110465240>, None, 21),
│   (<pickletools.OpcodeInfo object at 0x1104646a0>, 'type', 22),
│   (<pickletools.OpcodeInfo object at 0x110465240>, None, 28),
│   (<pickletools.OpcodeInfo object at 0x110465420>, None, 29),
│   (<pickletools.OpcodeInfo object at 0x110465240>, None, 30),
│   (<pickletools.OpcodeInfo object at 0x110464520>, None, 31),
│   (<pickletools.OpcodeInfo object at 0x110464ac0>, None, 32),
│   (<pickletools.OpcodeInfo object at 0x110465240>, None, 33),
│   (<pickletools.OpcodeInfo object at 0x110465480>, None, 34),
│   (<pickletools.OpcodeInfo object at 0x110465240>, None, 35),
│   (<pickletools.OpcodeInfo object at 0x110465720>, None, 36)
]

In [8]: pickletools.dis(pickle.dumps(type(None)))
    0: \x80 PROTO      4
    2: \x95 FRAME      26
   11: \x8c SHORT_BINUNICODE 'builtins'
   21: \x94 MEMOIZE    (as 0)
   22: \x8c SHORT_BINUNICODE 'type'
   28: \x94 MEMOIZE    (as 1)
   29: \x93 STACK_GLOBAL
   30: \x94 MEMOIZE    (as 2)
   31: N    NONE
   32: \x85 TUPLE1
   33: \x94 MEMOIZE    (as 3)
   34: R    REDUCE
   35: \x94 MEMOIZE    (as 4)
   36: .    STOP
highest protocol among opcodes = 4
```

Maybe we can do:

```diff
    def save_type(self, obj):
        if obj is type(None):
            return self.save_reduce(type, (None,), obj=obj)
        elif obj is type(NotImplemented):
            return self.save_reduce(type, (NotImplemented,), obj=obj)
        elif obj is type(...):
            return self.save_reduce(type, (...,), obj=obj)
+       elif obj.__module__ == 'functools':
+           import functools
+           if obj is type(functools.Placeholder):
+               return self.save_reduce(type, (functools.Placeholder,), obj=obj)
        return self.save_global(obj)
```

Reply:
> And what about `_pickle.c`?

https://github.com/python/cpython/blob/ead676516d4687b5e6d947a5c0808b5b93ba87e3/Modules/_pickle.c#L3851-L3864

You can fallback to my implementation in the original post:

```python
        elif obj.__module__ == 'functools':
            import functools  # avoid unnecessary import
            if obj is type(functools.Placeholder):
                return self.save_reduce(type, (functools.Placeholder,), obj=obj)
```

Reply:
Ok, so this approach seems to be more sensible than I initially thought.

Do you think this is better than exposing `SentinelType`?

Reply:
> Do you think this is better than exposing `SentinelType`?

I vote for not exposing that since the types of `None` / `Ellipsis` / `NotImplemented` are exposed in `types` rather than `builtins`.

Personally, I'd like to have a generic solution of sentinal variables. But that is kinda out of the scope of this PR. See [PEP 661 – Sentinel Values](https://peps.python.org/pep-0661).

```python
Placeholder = Sentinal('Placeholder', repr='<Placeholder>', module_name='functools'))
```

Reply:
Same, I had this thought reoccurring throughout this PR.

Also, have been battling with my own sentinels for quite some time. I think I have "good enough" solution for my own use cases, but I think standardisation of standard library sentinels will need to be addressed sooner or later.

This pickling issue was the last drop so I left a comment there: https://discuss.python.org/t/pep-661-sentinel-values/9126/133.

Much thanks for your help here. Would have taken me much longer to go through on my own.

Reply:
Also, this case of dual implementation (and Multi-phase module initialization) makes it hard to add `pickle` tests for sentinel object. This is the reason why this broken serialization has not been caught.

This would not be an issue if there was a dedicated sentinel factory - both C and Python would then just use the same sentinel.

Reply:
I have added all of the suggestions you made except `__repr__`. I think `__repr__` in sync with `None` is more appropriate here.

Reply:
I tested and it works for both `Placeholder` and `PlaceholderType` for both `C` and `Python` implementations for both `pickle.py` and `_pickle.c`.